### ISSUE TITLE
Questions about the performance in zero-shot pick-and-place task

### ISSUE THREAD

**Issue by Junxix** [vv19/instant_policy#1]
Hi! First of all, thank you for your great work!

We conducted a zero-shot real-world evaluation on a pick-and-place task (pick a cup, and place it into the large cup), using 9 demonstration samples as context. The observation is processed with SAM2. We have carefully checked the segmentation results, ensuring clear segmentation masks in both demonstrations and real-time observations.

![output_43](https://github.com/user-attachments/assets/e527fff7-b353-484e-895b-79067628543c)

The predicted action curve (using the end-effector TCP + Euler angles) appears to differ from the ground-truth action, especially in the z-axis. In our real robot evaluations, we also observe the same phenomenon: the robot tends to move away from the object (negative action z value) instead of approaching the object (positive action z value). Also, unlike the demo videos, the robot moves very slowly during deployment. 

![test](https://github.com/user-attachments/assets/48e1fe10-0006-49f6-9e47-e5d0edc18798)

We would greatly appreciate your insights on what might be causing this performance discrepancy.

**Comment by vv19** [comment]
Thanks a lot for trying it out!

Regarding the slow movement, it is all based on the controller that you use. We predict 8 actions as end-effector displacements and how fast you execute them is based on what controller you use (and how often you re-evaluate the policy).

Regarding the incorrect z value predictions, it is interesting. I would assume there might be some inconsistencies in coordinate frame representations. If you could share the data sample, (segmented point clouds + end-effector poses) I would be happy to investigate more.

**Comment by Junxix** [comment]
Thank you for your prompt response! We will experiment with different controllers as suggested. 

Additionally, we’ve uploaded 10 teleoperated demonstration data to Google Drive: [sample_dataset](https://drive.google.com/drive/folders/1kiOZqnJMNG9_bVvAA_gkuYdr32MEiUmZ). The demonstrations 1-9 are used as the context and we conducted offline evaluation using demonstration 10. Each demonstration is stored in a single .npy file with a format of `{
        "obs": {
            "pcds": pcds,
            "T_w_es": T_w_es,
            "grips": grips
        },
        "actions": {
            "actions": actions,
            "grips": action_grippers
        }
    }
`, and you can load them using `np.load("[filename].npy", allow_pickle=True).item()`. 

We truly appreciate your assistance!

**Comment by vv19** [comment]
Thanks! I'll look into it and will let you know how it goes. 

Just to clarify, when you tried rolling out a policy on the robot, it didn't work at all?

**Comment by Junxix** [comment]
Thank you for your response. When we attempted to roll out the policy in real-world experiments, we observed that due to the incorrect z-axis direction in our coordinate system transformation, the robotic arm and gripper unexpectedly moved upward instead of following the intended trajectory, which prevented successful task completion.  Below is the coordinate system of the robot end effector, where the X-axis of the end effector points into the image.

![WechatIMG7894](https://github.com/user-attachments/assets/84a76112-f956-43bd-b037-0b781cbd9926)

**Comment by vv19** [comment]
Overall, everything seems good to me (including the coordinate system). The predicted actions seem good, and when I visualise the predicted trajectories, they should be able to complete the task. So, I think the problem is how you convert predicted actions to robot commands.

The model outputs actions as (actions: [8, 4, 4], grips: [8, 1]), where actions are relative transformations of the end-effector. I assume you send the desired end-effector poses in the base frame to your controller. To compute desired end-effector poses from predicted actions, you need to do [T_w_e @ actions[j] for j in range(8)]. Here T_w_e: [4, 4] is the end-effector pose in the world (or base) frame. Doing this before you send commands to your controller should solve the problem. 

To avoid gripper actions from switching between -1 and 1, re-evaluate the policy when the gripper state changes, before sending new commands to your controller.

Other observations about your demonstrations:
- Your segmentations can cause a bit of an issue (see image below). I would suggest eroding the segmentation masks a little bit to avoid overflow. This can easily be done using cv2.erode(). 
- 'task_0516_user_0100_scene_0001_cfg_0001.npy' first moves towards one cup and then towards the other. This makes the extracted waypoints in the context non-consistent. When collecting demos avoid task irrelevant movements and try to make them consistent.
- To be better "in-distribution" with the data that the model was trained on, start recording the demonstrations with the end-effector ~30-50 cm above the objects and the gripper open and try to ensure that the trajectories are somewhat smooth.

![Screenshot from 2024-12-06 15-17-17](https://github.com/user-attachments/assets/c1c52c49-b1ff-46ca-b273-278ea4b5836a)

So, I would suggest first, adding the eroding of segmentation masks, re-recording a couple of demonstrations (e.g. 2) with the start pose of the end-effector a bit above the objects, correctly converting predicted actions to desired poses and rolling out the policy to see how it looks.

Let me know how it goes!

**Comment by Junxix** [comment]
Hi, thank you for your suggestions!

We checked the real-world deployment implementation and found bugs during the coordinate transformation parts. The bugs are fixed and now everything looks well. We also execute multiple steps during deployment to make the robot move faster. After every gripper state change, we re-evaluate the policies. We haven't collected new demonstration data (data with the start pose of the end-effector a bit above the objects) yet, but we will try it later. We use 10 demonstrations as contexts for real-world rollouts. 

Now the robot can make reasonable actions, _e.g._, approaching the cup and trying to grasp the cup. It can successfully grasp the cup a couple of times, especially when the end-effector is just above the cup. However, for other positions, the robot will approach the cup but can hardly grasp the cup. During successful grasps, the robot can lift the cup but cannot successfully move to the target area (the larger cup) and place it.

I believe more demonstrations will be needed to fine-tune the policy to adapt to the new scenario when the end-effector is not located a bit above the object. I will update more results (collecting new "in-distribution" demonstrations, trying to fine-tune the policy for the new scenario, _etc._) once finishing the experiments.

**Comment by Junxix** [comment]
By the way, since we are going to finetune the policy, how many demonstrations are needed for fine-tuning the policy on a single task?

**Comment by vv19** [comment]
Glad it "somewhat works now"! and thanks again for trying it out!

It is still strange to me that it doesn't complete this task. I just recreated your set-up (one depth camera, demonstrations start with the gripper between objects) and it worked very well for me. I used 2 demonstrations as context and that was enough. It might be the quality of the demonstrations you recorded. Try to ensure they are consistent and smooth. Also, executing all 8 predicted actions might help as well.

Regarding fine-tuning. As always, the more demos you have the better. In my opinion, 10 demos you have should definitely be enough. Create permutations of 2 demos for context, 1 for training and there are going to be quite a few samples already.

**Comment by vv19** [comment]
Also, when this happens "... the robot can lift the cup but cannot successfully move to the target area (the larger cup) and place it.", make sure that the camera actually sees the other cup and can segment it. With a single camera setup, it often happens that the other cup is occluded and therefore the policy can't move towards it.

**Comment by Junxix** [comment]
Thank you for your help! We executed all 8 predicted actions, except in cases where the gripper state changed. In such instances, we re-inferenced the policy to generate new action predictions. We also checked the camera and SAM2 results during the policy rollouts, and both appeared fine—there were no occlusions, and the segmentation was clear.

To clarify, when the robot successfully lifts the cup, it attempts to move towards the target. However, during this movement, the gripper state changes from "close" to "open," causing the object to drop. Based on the action prediction figure above, I hypothesize that the gripper state is not very stable during certain stages of the task.

Additionally, I would like to know whether your re-creation involves performing a similar task (e.g., picking up a cup and placing it in a larger cup) or if it replicates the original tasks described in the paper but with the gripper starting between objects?

Our demonstration data is collected through human teleoperation, which inevitably introduces some noise. To address this, we have carefully selected demonstrations from our dataset, aiming to ensure smooth context trajectories.

One notable aspect is that all our demonstrations begin with the gripper in a closed state. The robot moves to the top of the object, opens the gripper, approaches, and grasps the object. I plan to collect additional demonstrations where the gripper starts in an open state. I hypothesize that this gripper initial state inconsistency with your pre-trained data might cause the observed instability of the gripper.

**Comment by vv19** [comment]
Also, I just checked your demos again and noticed that your gripper states in the recorded demos are continuous values. They should be binary, either [0 or 1]. This alone might solve the problem.

**Comment by Junxix** [comment]
Oh these are the raw gripper data, we will process them to binary values before sending them into the policy, as shown in the "ground truth gripper action" figure in the first thread.

**Comment by vv19** [comment]
Nice. Just to clarify, the gripper states are assumed to be {0, 1} and save_sample(full_sample, None) on line 57 in deployment.py converts them to {-1, 1} before inputting them into the model. And the model outputs {-1, 1} as well.

**Comment by Junxix** [comment]
We collected 3 demonstrations using the same gripper state as pre-trained data (initial open, move to object and then close, move to target and then open). Using these demonstrations as context, this time, the policy works! It could successfully grasp the object, move to the target area, and place it a couple of times, especially when the object is under the gripper. We also observe that when the current gripper position is near the larger cup, the policy tends to grasp the larger cup instead of the desired object (small cup).  I hypothesize that both cups look quite similar in geometries, and without RGB it is hard to distinguish the small cup from the larger cup.

Nonetheless, we are excited that the policy can work zero-shot in a completely new environment with few-shot contexts. Thank you again for your help!

**Comment by vv19** [comment]
That is great to hear, well done! Indeed, without colour information, cups can look very similar and the model can confuse them. Colour and other semantic features could be added in the future to avoid these issues.

Would you mind sharing a video or a gif of the policy working? Would love to see it in action!
